site/content/entry/yuzu-fastmem/index.md

```diff
@@ -0,0 +1,118 @@
+
+To emulate this, we allocated a HUGE page table, which had an offset for each memory page addressable in the virtual memory region of the Switch. Just this table was 1GB! 
+And whenever a game wanted to access the memory, we looked up this table, which resulted in roughly ~10 x64 CPU instructions and twice the memory latency: First read the table's value, then read the correct memory.
+
+Fastmem, however, uses the host MMU to rebuild the same 39bit virtual memory arena. 
+Here we're lucky that Nintendo limits the memory to 39bit. 
+We cannot allocate 48bit because that's the entire virtual address space on the host. 
+
+This cuts the effort to access the memory down to 3+1 x64 CPU instructions with a single memory latency.
+3 instructions to check if the pointer is smaller than 39 bit, and one memory load instruction. 
+We need this size check to ensure out-of-bounds memory access won't return any physical address.
+
+
+## The Challenges
+
+Implementing Fastmem required two major things on [Dynarmic](https://github.com/MerryMage/dynarmic/), our JIT recompiler. 
+
+**First:** A64 Fastmem support on Dynarmic.
+
+Dynarmic has had support for A32 Fastmem (see Citra Android) for a while now. 
+A64 Fastmem support was [originally worked on a year ago by MerryMage and vdwjeremy](https://github.com/MerryMage/dynarmic/pull/528) but it never got merged to the master branch. 
+
+[degasus](https://github.com/degasus/) worked with [MerryMage](https://github.com/MerryMage/) to clean up and get these changes merged to the master branch so that yuzu could benefit from this. [See here](https://github.com/MerryMage/dynarmic/pull/613)
+
+**Second:** To generate the 512GB virtual address space with the same mappings as on the Switch.
+
+degasus proved last year that this was easy to implement on Linux as the POSIX mmap call provided very similar features to the Horizon OS.
+But it was a huge challenge for Windows, as the Windows Virtual Memory API is larger with a few missing features. 
+Moreover, while Windows page tables were 4K aligned, Windows memory management was 64K aligned for outdated reasons. [See here](https://devblogs.microsoft.com/oldnewthing/20031008-00/?p=42223)
+
+
+## The Solutions
+
+Turns out, Microsoft had realized this too and frequently patched their Virtual Memory API. 
+With the release of Windows 10 v1803, their new API supported new "placeholder" memory which didn't suffer from the previous limitations. 
+
+Thanks to [Breadfish64](https://github.com/BreadFish64/), who used this new API in their Gameboy emulator and showed us that it was now indeed possible to support Fastmem on Windows. 
+Unfortunately, this makes the feature usable only on Windows versions 1803 and higher.
+
+Since we already met the base requirements for our new fastmem implementation long ago, our devs were able to quickly address the missing pieces. 
```

Comment:
```suggestion
Since we already met the base requirements for our new Fastmem implementation long ago, our devs were able to quickly address the missing pieces. 
```